crates/core/src/network/io/reactor.rs

```diff
@@ -0,0 +1,465 @@
+                    ),
+                )
+                .await
+                {
+                    warn!(
+                        target = target.as_str_name(),
+                        "Error processing message: {e}"
+                    );
+                }
+                histogram!(NETWORK_MESSAGE_PROCESSING_DURATION, "target" => target.as_str_name())
+                    .record(processing_started.elapsed());
+                counter!(NETWORK_MESSAGE_RECEIVED, "target" => target.as_str_name()).increment(1);
+                counter!(NETWORK_MESSAGE_RECEIVED_BYTES, "target" => target.as_str_name())
+                    .increment(encoded_len as u64);
+                trace!(
+                    target = target.as_str_name(),
+                    "Processed message in {:?}",
+                    processing_started.elapsed()
+                );
+                Decision::Continue
+            }
+        }
+    }
+}
+
+#[derive(derive_more::Index, derive_more::IndexMut)]
+pub struct MetadataVersions {
+    #[index_mut]
+    #[index]
+    versions: EnumMap<MetadataKind, Version>,
+    metadata: Metadata,
+    nodes_config: Live<NodesConfiguration>,
+    schema: Live<Schema>,
+    partition_table: Live<PartitionTable>,
+    logs_metadata: Live<Logs>,
+}
+
+impl MetadataVersions {
+    fn new(metadata: Metadata) -> Self {
+        let mut nodes_config = metadata.updateable_nodes_config();
+        let mut schema = metadata.updateable_schema();
+        let mut partition_table = metadata.updateable_partition_table();
+        let mut logs_metadata = metadata.updateable_logs_metadata();
+
+        let versions = enum_map! {
+            MetadataKind::NodesConfiguration => nodes_config.live_load().version(),
+            MetadataKind::Schema => schema.live_load().version(),
+            MetadataKind::PartitionTable => partition_table.live_load().version(),
+            MetadataKind::Logs => logs_metadata.live_load().version(),
+        };
+
+        Self {
+            metadata,
+            versions,
+            nodes_config,
+            schema,
+            partition_table,
+            logs_metadata,
+        }
+    }
+
+    pub fn notify(&mut self, header: &Header, connection: &Connection) {
+        self.update(
+            header.my_nodes_config_version.map(Into::into),
+            header.my_partition_table_version.map(Into::into),
+            header.my_schema_version.map(Into::into),
+            header.my_logs_version.map(Into::into),
+        )
+        .into_iter()
+        .for_each(|(kind, version)| {
+            if let Some(version) = version {
+                if version > self.get_latest_version(kind) {
+                    self.metadata.notify_observed_version(
+                        kind,
+                        version,
+                        Some(connection.clone()),
+                        Urgency::Normal,
+                    );
+                }
+                // todo: store the latest if it's higher
```

Comment:
We are storing the versions in `self.versions`. Do you mean to store it somewhere else to avoid calling `notify_observed_version` multiple times?

Reply:
yes